src/Rules/Generate.hs

```diff
@@ -99,76 +100,92 @@ generate file context expr = do
-                    ++ fmap (root -/-) includesDependencies
-                build $ target context HsCpp [primopsSource] [file]
-
-            "//" ++ platformH stage %> go generateGhcBootPlatformH
-
-        -- TODO: why different folders for generated files?
-        priority 2.0 $ fmap (("//" ++ dir) -/-)
-            [ "GHC/Prim.hs"
-            , "GHC/PrimopWrappers.hs"
-            , "*.hs-incl" ] |%> \file -> do
-                root <- buildRoot
-                need [root -/- primopsTxt stage]
-                build $ target context GenPrimopCode [root -/- primopsTxt stage] [file]
-
-        when (pkg == rts) $ "//" ++ dir -/- "cmm/AutoApply.cmm" %> \file ->
-            build $ target context GenApply [] [file]
-
--- TODO: These rules copy runtime dependencies of some executables, such as GHC
--- itself (file @ghc-usage.txt@) or Hsc2Hs (file @template-hsc.h@). Ideally,
--- these rules should be moved to package-specific settings, so that they can be
--- discovered more easily. We also need to add proper support for runtime
--- dependencies on directories, which is the case for Haddock -- for the current
--- workaround see "Rules.Documentation.haddockHtmlResourcesStamp".
+    generated ?> \file -> do
+      let unpack = fromMaybe . error $ "No generator for " ++ file ++ "."
+      (src, builder) <- unpack <$> findGenerator context file
+      need [src]
+      build $ target context builder [src] [file]
+      let boot = src -<.> "hs-boot"
+      whenM (doesFileExist boot) . copyFile boot $ file -<.> "hs-boot"
+
+    priority 2.0 $ do
+        when (pkg == compiler) $ do root <//> dir -/- "Config.hs" %> go generateConfigHs
+                                    root <//> dir -/- "*.hs-incl" %> genPrimopCode context
+        when (pkg == ghcPrim) $ do (root <//> dir -/- "GHC/Prim.hs") %> genPrimopCode context
+                                   (root <//> dir -/- "GHC/PrimopWrappers.hs") %> genPrimopCode context
+        when (pkg == ghcPkg) $ do root <//> dir -/- "Version.hs" %> go generateVersionHs
+
+    -- TODO: needing platformH is ugly and fragile
+    when (pkg == compiler) $ do
+        root -/- primopsTxt stage %> \file -> do
+            root <- buildRoot
+            need $ [ root -/- platformH stage
+                   , primopsSource]
+                ++ fmap (root -/-) includesDependencies
+            build $ target context HsCpp [primopsSource] [file]
+
+        -- only generate this once! Until we have the include logic fixed.
+        -- See the note on `platformH`
+        when (stage == Stage0) $ do
+           root <//> "compiler/ghc_boot_platform.h" %> go generateGhcBootPlatformH
+        root <//> platformH stage %> go generateGhcBootPlatformH
+
+    when (pkg == rts) $ do
+      root <//> dir -/- "cmm/AutoApply.cmm" %> \file ->
+        build $ target context GenApply [] [file]
+
+      -- XXX: this should be fixed properly, e.g. generated here on demand.
```

Comment:
Could you expand this comment? It is unclear to me what exactly needs to be done.

Reply:
We have this rule as well as `generateRules`. I think we should inline the`generateRules` logic here and get rid of the two rule process. With that we can probably get rid of the `generatedDir` as well.

Reply:
@angerman Sorry, I still don't understand :) Perhaps, you could raise an issue where this could be discussed further? We'll just link this issue here.